pipeline-runner/R/gem2s-4-score_doublets.R

```diff
@@ -1,68 +1,65 @@
-  message("reloading old matrices...")
-
-  scdata_list <- readRDS("/output/pre-doublet-scdata_list.rds")
-
-  message("Loading configuration...")
-  config <- RJSONIO::fromJSON("/input/meta.json")
-
-  print_config(4, "Score Doublets", input, pipeline_config, config)
-
-  # Check which samples have been selected. Otherwiser we are going to use all of them.
-  if (length(config$samples) > 0) {
-    samples <- config$samples
-  } else {
-    samples <- names(scdata_list)
-  }
-
-  scdata_list <- scdata_list[samples]
-
-  message("calculating probability of barcodes being doublets...")
-  for (sample_name in names(scdata_list)) {
-    compute_doublet_scores(scdata_list[[sample_name]], sample_name)
+#' Compute doublet score for count matrices
+#'
+#' @inheritParams download_cellranger
+#' @param prev_out 'output' slot from call to \code{run_emptydrops}
+#'
+#' @return \code{prev_out} with added slot 'doublet_scores' containing result of call to
+#'   \code{\link{compute_doublet_scores}} for each sample.
+#'
+#' @export
+#'
+score_doublets <- function(input, pipeline_config, prev_out) {
+  message("Calculating probability of droplets being doublets...")
+
+  edrops_list <- prev_out$edrops
+  counts_list <- prev_out$counts_list
+  samples <- names(counts_list)
+
+  scores <- list()
+  for (sample in samples) {
+    message("\nSample --> ", sample)
+    counts <- counts_list[[sample]]
+    edrops <- edrops_list[[sample]]
+
+    # scDblFinder expects empty droplets to be removed
+    if (!is.null(edrops)) {
+      keep <- which(edrops$FDR <= gem2s$max.edrops.fdr)
+      counts <- counts[, keep]
+    }
+
+    # also filter low UMI as per scDblFinder:::.checkSCE()
+    ntot <- Matrix::colSums(counts)
+    counts <- counts[, ntot > 200]
+
+    scores[[sample]] <- compute_doublet_scores(counts)
   }
 
-  message("Step 4 completed.")
+  prev_out$doublet_scores <- scores
+  res <- list(
+    data = list(),
+    output = prev_out)
 
-  return(list())
+  message("\nScoring doublets step complete.")
+  return(res)
 }
 
 
-# compute_doublet_scores function
-#' @description Save the result of doublets scores per sample.
+#' Compute doublets scores per sample.
+#'
 #' @param scdata Raw sparse matrix with the counts for one sample.
-#' @param sample_name Name of the sample that we are preparing.
 #'
-#' @return
-compute_doublet_scores <- function(scdata, sample_name) {
-  message("loading scd")
-  library(scDblFinder)
-  message("loaded")
-  message("Sample --> ", sample_name, "...")
-
-
-  edpath <- paste0("/output/pre-emptydrops-", sample_name, ".rds")
-  if (file.exists(edpath)) {
-    edout <- readRDS(edpath)
-    keep <- which(edout$FDR <= 0.001)
-    scdata <- scdata[, keep]
-  }
-
-  # also filter low UMI as per scDblFinder:::.checkSCE()
-  ntot <- Matrix::colSums(scdata)
-  scdata <- scdata[, ntot > 200]
+#' @return data.frame with doublet scores and assigned classes
+#'
+compute_doublet_scores <- function(counts) {
```

Comment:
rename this to `compute_sample_doublet_scores`, so that it is obvious that this is per sample, rename `counts` to `sample_counts` to make it obvious that this is for a sample.